### ISSUE TITLE
need mechanism to render training accuracies highly deterministic for testing

### ISSUE THREAD

**Issue by beatthat** [opentutor/opentutor-classifier#42]
Currently it seems that tests that use fixed training data and config can nonetheless get some variance in training accuracy.

This forces us to put undesirably large epsilons on test assertions about training accuracy. It would be better if we could make accuracy metrics fully deterministic, at least for the scope of running tests. We don't currently know for sure what causes the non determinism, but best current guess is something random inside scikit learn with respect to sectioning training vs cross-validation data. 
<img width="1249" alt="Screen Shot 2020-11-30 at 5 12 04 PM" src="https://user-images.githubusercontent.com/323901/100684782-d8e07400-332f-11eb-8db3-9ff7c6568175.png">

**Comment by markcore** [comment]
This was triggered by a new test example in branch classifier-improvements: the test is test_train_online and the new example is "ies-television". The accuraries were not consistent and sometimes the old test framework (using precision) would fail.

**Comment by markcore** [comment]
One of the assertions is checking the accuracy recorded during training:
```
            results_loocv = model_selection.cross_val_score(
                model, features, train_y, cv=LeaveOneOut(), scoring="accuracy"
            )
```
Each sample is used once as a test set (singleton) while the remaining samples form the training set.

**Comment by markcore** [comment]
It seems like the results vary by one example being right/wrong. I'm also noticing slight differences in the word2vec-similarity scores:
EXAMPLE     FEATURES
['new', 'cost', 'less', 'original', 'price', 'percentages', 'ratios', 'numbers', 'would', 'multiply', 'everything', 'together']	[0.0, 0.0, 0.0, 1.0, 0.5467495530843735, 2.4, 0.6813951134681702, **0.8156701922416687**]
 ['new', 'cost', 'less', 'original', 'price', 'percentages', 'ratios', 'numbers', 'would', 'multiply', 'everything', 'together']	[0.0, 0.0, 0.0, 1.0, 0.5467495530843735, 2.4, 0.6813951134681702, **0.8156702518463135**]

**Comment by markcore** [comment]
This is happening in features.py. A method called _avg_feature_vector calculates the vectors, and spatial.distance.cosine takes those and does the work of calculating the similarity. It isn't clear why any of those would be different between runs.

**Comment by markcore** [comment]
Looks like it is vector representation that changes for some reason from run to run.

**Comment by markcore** [comment]
I pushed two commits into a new branch. The accuracy is only off by 1% between runs so I tightened epsilon; also, the accuracy in the test was off.

There are a lot of floating point operations. I'm not sure why representing the decimals would change from run to run but as best I can tell that's what is happening.

**Comment by beatthat** [comment]
resolved. we now generally run tests against a test data set and expect an overall accuracy
